### PR TITLE
Refactor git command arguments and make all arguments to be safe to be used

### PR DESCRIPTION
**Pull request by wxiaoguang**
Follow
* https://github.com/go-gitea/gitea/pull/21464

Purpose of this PR:
* Make all git command arguments strictly safe

Non-purpose:
* To be backported
* Change any existing logic

### CODE REVIEW

**Comment by delvh** [comment]
But aren't we actually regressing a lot in this PR? 
There were a lot of places that previously used
```sh
git $COMMAND -- $WHATEVER
```
that are now refactored to
```sh
git $COMMAND $(if [[ "$WHATEVER" == "-"* ]] ; then exit 1; else echo "$WHATEVER"; fi)
```
so we simply don't allow something that was previously possible and worked correctly?
That is the definition of highly-**breaking**…

**Comment by wxiaoguang** [comment]
How  could it break? Please show the exactly example.

The `-- ...` arguments are all replaced by `AddSlashedArguments`,  which handles the `--` correctly.

**Comment by delvh** [comment]
Ah right, seems as if I was too sleepy when I wrote that to realize that the calls weren't `AddDynamicArguments` but `AddSlashedArguments` instead.
Sorry for the confusion!

**Comment by lunny** [comment]
Why the changes to `AddSlashedArguments` is necessary?

**Comment by wxiaoguang** [comment]
> Why the changes to `AddSlashedArguments` is necessary?

Why not necessary?

**Comment by wxiaoguang** [comment]
@go-gitea/maintainers 

Need reviews. If no interest, I will close it 

Although I have proposed this solution since Apr, but I do not feel I got enough supports for it, this is my last try for this problem, if this proposal doesn't get accepted, please use other approaches to reinforce the git command system, otherwise there will be more argument injection or remote code execution problems.

**Comment by delvh** [comment]
I'm on it.
It simply takes quite a while due to the amount of changes.

File: services/pull/temp_repo.go [hunk 1]
```diff
@@ -124,7 +124,7 @@ func createTemporaryRepo(ctx context.Context, pr *issues_model.PullRequest) (str
 	outbuf.Reset()
 	errbuf.Reset()
 
-	if err := git.NewCommand(ctx, "symbolic-ref", "HEAD", git.BranchPrefix+baseBranch).
+	if err := git.NewCommand(ctx, "symbolic-ref").AddDynamicArguments("HEAD", git.BranchPrefix+baseBranch).
```
Comment (delvh) [inline services/pull/temp_repo.go:127]:
```suggestion
	if err := git.NewCommand(ctx, "symbolic-ref", "HEAD").AddDynamicArguments(git.BranchPrefix+baseBranch).
```

File: modules/git/repo_blame.go [hunk 2]
```diff
@@ -8,13 +8,16 @@ import "fmt"
...
+	stdout, _, err := NewCommand(repo.Ctx, "blame", "--root").AddSlashedArguments(file).RunStdBytes(&RunOpts{Dir: path})
 	return stdout, err
 }
 
 // LineBlame returns the latest commit at the given line
 func (repo *Repository) LineBlame(revision, path, file string, line uint) (*Commit, error) {
-	res, _, err := NewCommand(repo.Ctx, "blame", fmt.Sprintf("-L %d,%d", line, line), "-p", revision, "--", file).RunStdString(&RunOpts{Dir: path})
+	res, _, err := NewCommand(repo.Ctx, "blame").
+		AddArguments(CmdArg(fmt.Sprintf("-L %d,%d", line, line))).
+		AddArguments("-p").AddDynamicArguments(revision).
```
Comment (delvh) [inline modules/git/repo_blame.go:19]:
```suggestion
	res, _, err := NewCommand(repo.Ctx, "blame", CmdArg(fmt.Sprintf("-L %d,%d", line, line)), "-p").
		AddDynamicArguments(revision).
```

File: modules/git/repo_commit.go [hunk 3]
```diff
@@ -89,8 +89,10 @@ func (repo *Repository) GetCommitByPath(relpath string) (*Commit, error) {
 }
 
 func (repo *Repository) commitsByRange(id SHA1, page, pageSize int) ([]*Commit, error) {
-	stdout, _, err := NewCommand(repo.Ctx, "log", id.String(), "--skip="+strconv.Itoa((page-1)*pageSize),
-		"--max-count="+strconv.Itoa(pageSize), prettyLogFormat).RunStdBytes(&RunOpts{Dir: repo.Path})
+	stdout, _, err := NewCommand(repo.Ctx, "log").
+		AddArguments(CmdArg("--skip="+strconv.Itoa((page-1)*pageSize)), CmdArg("--max-count="+strconv.Itoa(pageSize)), prettyLogFormat).
```
Comment (delvh) [inline modules/git/repo_commit.go:93]:
```suggestion
	stdout, _, err := NewCommand(repo.Ctx, "log", CmdArg("--skip="+strconv.Itoa((page-1)*pageSize)), CmdArg("--max-count="+strconv.Itoa(pageSize)), prettyLogFormat).
```

File: modules/repository/init.go [hunk 4]
```diff
@@ -317,41 +317,41 @@ func initRepoCommit(ctx context.Context, tmpPath string, repo *repo_model.Reposi
...
+		log.Error("Failed to commit: %v: Stdout: %s\nError: %v", cmd.String(), stdout, err)
 		return fmt.Errorf("git commit: %v", err)
 	}
 
 	if len(defaultBranch) == 0 {
 		defaultBranch = setting.Repository.DefaultBranch
 	}
 
-	if stdout, _, err := git.NewCommand(ctx, "push", "origin", "HEAD:"+defaultBranch).
+	if stdout, _, err := git.NewCommand(ctx, "push", "origin").AddDynamicArguments("HEAD:" + defaultBranch).
```
Comment (delvh) [inline modules/repository/init.go:354]: That is not dynamic?

**Comment by wxiaoguang** [comment]
Actually there are not too many changes. The new mechanism guarantees there is no low-level mistake (unless anything is lost by mistake).

* git command accepts `CmdArg` , which is considered to be safe.
* for unclear arguments, use `AddDynamicArguments`
* for file/ref list, use `AddSlashedArguments`

File: services/migrations/gitea_uploader_test.go [hunk 5]
```diff
@@ -234,7 +234,7 @@ func TestGiteaUploadUpdateGitForPullRequest(t *testing.T) {
 	fromRepo := unittest.AssertExistsAndLoadBean(t, &repo_model.Repository{ID: 1})
 	baseRef := "master"
 	assert.NoError(t, git.InitRepository(git.DefaultContext, fromRepo.RepoPath(), false))
-	err := git.NewCommand(git.DefaultContext, "symbolic-ref", "HEAD", git.BranchPrefix+baseRef).Run(&git.RunOpts{Dir: fromRepo.RepoPath()})
+	err := git.NewCommand(git.DefaultContext, "symbolic-ref").AddDynamicArguments("HEAD", git.BranchPrefix+baseRef).Run(&git.RunOpts{Dir: fromRepo.RepoPath()})
```
Comment (delvh) [inline services/migrations/gitea_uploader_test.go:237]:
```suggestion
	err := git.NewCommand(git.DefaultContext, "symbolic-ref", "HEAD").AddDynamicArguments(git.BranchPrefix+baseRef).Run(&git.RunOpts{Dir: fromRepo.RepoPath()})
```

File: services/pull/merge.go [hunk 6]
```diff
@@ -244,7 +244,7 @@ func rawMerge(ctx context.Context, pr *issues_model.PullRequest, doer *user_mode
 	stagingBranch := "staging"
 
 	if expectedHeadCommitID != "" {
-		trackingCommitID, _, err := git.NewCommand(ctx, "show-ref", "--hash", git.BranchPrefix+trackingBranch).RunStdString(&git.RunOpts{Dir: tmpBasePath})
+		trackingCommitID, _, err := git.NewCommand(ctx, "show-ref", "--hash").AddDynamicArguments(git.BranchPrefix + trackingBranch).RunStdString(&git.RunOpts{Dir: tmpBasePath})
```
Comment (delvh) [inline services/pull/merge.go:247]: dynamic? But can't we trust `BranchPrefix`?

File: services/pull/merge.go [hunk 7]
```diff
@@ -643,7 +646,7 @@ func commitAndSignNoAuthor(ctx context.Context, pr *issues_model.PullRequest, me
 			return fmt.Errorf("git commit [%s:%s -> %s:%s]: %v\n%s\n%s", pr.HeadRepo.FullName(), pr.HeadBranch, pr.BaseRepo.FullName(), pr.BaseBranch, err, outbuf.String(), errbuf.String())
 		}
 	} else {
-		if err := git.NewCommand(ctx, "commit", signArg, "-m", message).
+		if err := git.NewCommand(ctx, "commit").AddArguments(signArg).AddArguments("-m").AddDynamicArguments(message).
```
Comment (delvh) [inline services/pull/merge.go:649]: We trust the signArg?

File: services/pull/temp_repo.go (hunk 1 above)
Comment (wxiaoguang) [reply services/pull/temp_repo.go:127]: No need.

File: modules/git/repo_blame.go (hunk 2 above)
Comment (wxiaoguang) [reply modules/git/repo_blame.go:19]: No need

File: modules/git/repo_commit.go (hunk 3 above)
Comment (wxiaoguang) [reply modules/git/repo_commit.go:93]: It's too long in one time

File: modules/repository/init.go (hunk 4 above)
Comment (wxiaoguang) [reply modules/repository/init.go:354]: It is

File: services/migrations/gitea_uploader_test.go (hunk 5 above)
Comment (wxiaoguang) [reply services/migrations/gitea_uploader_test.go:237]: No need

File: services/pull/merge.go (hunk 6 above)
Comment (wxiaoguang) [reply services/pull/merge.go:247]: It must be written as this.

File: services/pull/merge.go (hunk 7 above)
Comment (wxiaoguang) [reply services/pull/merge.go:649]: Yes, it is `CmdArg`

File: modules/repository/init.go (hunk 4 above)
Comment (delvh) [reply modules/repository/init.go:354]: Under the current assumptions, how can this be an invalid command?
It always starts with `H` and not with `-`…

**Comment by wxiaoguang** [comment]
To explain more about `AddDynamicArguments`:

It doesn't have side effect if the argument is safe. Any non-option argument could be added by it.

And it must be used if the argument is a string (eg: `BranchPrefix + name`).

File: services/pull/temp_repo.go (hunk 1 above)
Comment (wxiaoguang) [reply services/pull/temp_repo.go:127]: `git symbolic-ref [-m <reason>] <name> <ref>`

So the `name` and `ref` could all be dynamic arguments, so put them together.

File: modules/repository/init.go (hunk 4 above)
Comment (wxiaoguang) [reply modules/repository/init.go:354]: See my comment
* https://github.com/go-gitea/gitea/pull/21535#issuecomment-1288087921

**Review by delvh — APPROVED** [review]

File: models/migrations/v128.go [hunk 8]
```diff
@@ -83,17 +83,17 @@ func fixMergeBase(x *xorm.Engine) error {
 
 			if !pr.HasMerged {
 				var err error
-				pr.MergeBase, _, err = git.NewCommand(git.DefaultContext, "merge-base", "--", pr.BaseBranch, gitRefName).RunStdString(&git.RunOpts{Dir: repoPath})
+				pr.MergeBase, _, err = git.NewCommand(git.DefaultContext, "merge-base").AddSlashedArguments(pr.BaseBranch, gitRefName).RunStdString(&git.RunOpts{Dir: repoPath})
 				if err != nil {
 					var err2 error
-					pr.MergeBase, _, err2 = git.NewCommand(git.DefaultContext, "rev-parse", git.BranchPrefix+pr.BaseBranch).RunStdString(&git.RunOpts{Dir: repoPath})
+					pr.MergeBase, _, err2 = git.NewCommand(git.DefaultContext, "rev-parse").AddDynamicArguments(git.BranchPrefix + pr.BaseBranch).RunStdString(&git.RunOpts{Dir: repoPath})
```
Comment (KN4CK3R) [inline models/migrations/v128.go]: Maybe a bit late to the party but for me the names `AddSlashedArguments` and `AddDynamicArguments` are not very descriptive. In the example above `pr.BaseBranch` is added one time with `AddSlashedArguments` and the other time with `AddDynamicArguments`. Without deep understanding what both methods do I can't tell the difference.

How about renaming `AddSlashedArguments` to `ForceEndOfOptionsAndAddArguments` (🎉) and `AddDynamicArguments` to `AddArgumentsAndPreventOptions` or `AddUnsafeArguments`. Feel free to propose better names.

Comment (wxiaoguang) [reply models/migrations/v128.go]: Because the code base is messy at the moment. Many arguments are not added by proper methods.

This PR's purpose is just making everything safe, but not changing any existing logic.

Actually I have tried `AddUnsafeArguments` before, but it's even more unclear .... because many dynamic arguments (user-provided or constructed by code)  are safe.


`ForceEndOfOptionsAndAddArguments` doesn't work either because there are two different options: `--end-of-options` and `--`, they are not the same.

Comment (wxiaoguang) [reply models/migrations/v128.go]: And you could even see other mistakes in old code, like this:

image

I am sure the old `--` was at the wrong place, but as a (big) refactoring, I do not want to touch the old logic. This PR is safe enough and should work well in all cases.

Leave more improvements to next PRs.

Comment (KN4CK3R) [reply models/migrations/v128.go]: > Because the code base is messy at the moment. Many arguments are not added by proper methods.
> 
> This PR's purpose is just making everything safe, but not changing any existing logic.

Changing the method names does not change existing logic.

> Actually I have tried `AddUnsafeArguments` before, but it's even more unclear .... because many dynamic arguments (user-provided or constructed by code) are safe.

How about `AddPossiblyUnsafeArguments`?

> `ForceEndOfOptionsAndAddArguments` doesn't work either because there are two different options: `--end-of-options` and `--`, they are not the same.

We don't use `--end-of-options` at the moment. GitLab calls arguments after `--` "post separator args".

And why `Slashed` if the method adds `--`? Those are `dashes` and not `slashes` (`/`).

Comment (wxiaoguang) [reply models/migrations/v128.go:89]: Oh yes .... I chosen the wrong word ....... then `AddDashedArguments`?

Comment (KN4CK3R) [reply models/migrations/v128.go]: To get this PR in, yes.

Comment (wxiaoguang) [reply models/migrations/v128.go]: > Changing the method names does not change existing logic.

I am answering "why the different methods are used to BaseBranch"  😂

> How about AddPossiblyUnsafeArguments?

It seems too wordy .....

> And why Slashed if the method adds --? Those are dashes and not slashes (/).

Oh yes .... I chosen the wrong word ....... then AddDashedArguments?

**Review by KN4CK3R — APPROVED** [review]
